File: src/main/java/com/kiborisaway/tasktimetracker/data/entity/Task.java
package com.kiborisaway.tasktimetracker.data.entity;

import com.kiborisaway.tasktimetracker.data.dto.task.TaskCreateRequest;
import com.kiborisaway.tasktimetracker.data.dto.task.TaskUpdatePropertyRequest;
import java.time.LocalDateTime;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@EqualsAndHashCode
@AllArgsConstructor
@NoArgsConstructor
public class Task {

  private Integer id;

  private Integer projectId;//taskGroupIdを持つならここは持たない

  private Integer taskGroupId;

  private String title;

  private String description;

  private Integer estimatedMinutes;

  private LocalDateTime createdAt;

  private LocalDateTime finishedAt;

  private Integer actualMinutesCached;

  private Integer gapMinutesCached;

  private Double gapRateCached;

  /**
   * タスク新規登録用。projectId・taskGroupIdはパスパラメータ由来のためDTOに含まれません。
   *
   * @param projectId   プロジェクトID
   * @param taskGroupId タスクグループID
   * @param request     タスク新規登録リクエスト
   */
  public Task(Integer projectId, Integer taskGroupId, TaskCreateRequest request) {
    this.projectId = projectId;
    this.taskGroupId = taskGroupId;
    this.title = request.getTitle();
    this.description = request.getDescription();
    this.estimatedMinutes = request.getEstimatedMinutes();
  }

  /**
   * タスク名・説明更新用。
   *
   * @param id      タスクID
   * @param request タスク名・説明更新リクエスト
   */
  public Task(int id, TaskUpdatePropertyRequest request) {
    this.id = id;
    this.title = request.getTitle();
    this.description = request.getDescription();
  }
}

File: frontend/src/types/task.ts
import type { MemoResponse } from './memo'
import type { TagSummary } from './tag'

export interface TaskResponse {
  id: string
  projectId: string | null
  taskGroupId: string | null
  title: string
  description: string | null
  estimatedMinutes: number | null
  createdAt: string
  finishedAt: string | null // ← 完了判定はこれ。isFinished は存在しない
  actualMinutesCached: number | null // 未完了時は null
  gapMinutesCached: number | null // 未完了時は null
  gapRateCached: number | null // 未完了時は null。単位は「％」（100倍しない）
  memos: MemoResponse[]
  tags: TagSummary[]
}

export interface TaskCreateRequest {
  title: string
  description: string | null
  estimatedMinutes: number
  tagIds: string[]
}

export interface TaskUpdatePropertyRequest {
  title: string
  description: string | null
}

export interface TaskUpdateEstimatedMinutesRequest {
  estimatedMinutes: number
}

export interface TaskUpdateFinishedRequest {
  isFinished: boolean
}

export interface TaskUpdateParentRequest {
  projectId: string | null
  taskGroupId: string | null
}

export interface TaskTagsUpdateRequest {
  tagIds: string[]
}
